src/tools/schema2graph.ts

```diff
@@ -80,59 +96,37 @@ export class SchemaNode {
+  // describes the reviews of Product read with the favProduct connection.
+  //
+  // Address names are the only names that can be relied on being available outside the scope of
+  // the particle due to generated type aliases.
   fullName(connection: HandleConnectionSpec): string {
-    return this.nearestSourceForConnection(connection).fullName;
-  }
-
-  private nearestSourceForConnection(connection: HandleConnectionSpec): SchemaSource {
-    const sourcesFromConnection = this.sources.filter(s => s.connection === connection);
-    const minPathLength = Math.min(...sourcesFromConnection.map(s => s.path.length));
-    return sourcesFromConnection.find(s => s.path.length === minPathLength);
-  }
-
-  // This method is a temporary workaround. To fully support tuples we need to enhance the spec
-  // definition for Kotlin handles.
-  // TODO(b/157598151): Update HandleSpec from hardcoded single EntitySpec to
-  //                    allowing multiple EntitySpecs for handles of tuples.
-  static singleSchemaHumanName(connection: HandleConnectionSpec, nodes: SchemaNode[]): string {
-    const topLevelNode = SchemaNode.findTopLevelNodes(connection, nodes)[0];
-    return topLevelNode.humanName(connection);
-  }
-
-  // This method is a temporary workaround. To fully support tuples we need to enhance the spec
-  // definition for Kotlin handles.
-  // TODO(b/157598151): Update HandleSpec from hardcoded single EntitySpec to
-  //                    allowing multiple EntitySpecs for handles of tuples.
-  static singleSchemaFullName(connection: HandleConnectionSpec, nodes: SchemaNode[]): string {
-    const topLevelNode = SchemaNode.findTopLevelNodes(connection, nodes)[0];
-    return topLevelNode.fullName(connection);
+    const sourcesForThisConnection = this.sources.filter(s => s.connection === connection);
+    const representativeSource = SchemaSource.filterToShortestPaths(sourcesForThisConnection)[0];
+    return representativeSource.fullName;
   }
 
   // Returns all "top-level" schema nodes for the given connection.
   // There will be a single one for handles of entities or references to entities,
-  // but arbitrary many for handles of tuples.
-  private static findTopLevelNodes(connection: HandleConnectionSpec, nodes: SchemaNode[]): SchemaNode[] {
-    const sourcesFromConnection = nodes
-        .map(n => n.sources)
-        .reduce((curr, acc) => [...acc, ...curr], [])
-        .filter(s => s.connection === connection);
-    const minPathLength = Math.min(...sourcesFromConnection.map(s => s.path.length));
-    const bestSources = sourcesFromConnection.filter(s => s.path.length === minPathLength);
-    return nodes.filter(n => n.sources.some(s => bestSources.includes(s)));
+  // but arbitrarily many for handles of tuples.
+  static topLevelNodes(connection: HandleConnectionSpec, nodes: SchemaNode[]): SchemaNode[] {
+    const allSources = nodes.map(n => n.sources) .reduce((curr, acc) => [...acc, ...curr], []);
```

Comment:
I added a `flatMap` method in `util.ts` 😉 

`const allSources = flatMap(nodes, n => n.sources);`

Reply:
Done :)